src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs

```diff
@@ -307,26 +307,40 @@ private static string GetRestoreProjectPath(IVsTargetFrameworks tfms)
             return GetNonEvaluatedPropertyOrNull(tfms, RestorePackagesPath, e => e);
         }
 
+        /// <summary>
+        /// The result will contain CLEAR and no sources specified in RestoreSources if the clear keyword is in it.
+        /// If there are additional sources specified, the value RestoreAdditionalProjectSources will be set in the result and then all the additional sources will follow
+        /// </summary>
         private static string[] GetRestoreSources(IVsTargetFrameworks tfms)
         {
             var sources = MSBuildStringUtility.Split(GetNonEvaluatedPropertyOrNull(tfms, RestoreSources, e => e));
 
             sources = HandleClear(sources);
 
             var additional = MSBuildStringUtility.Split(GetNonEvaluatedPropertyOrNull(tfms, RestoreAdditionalProjectSources, e => e));
-            sources = sources.Concat(additional).ToArray();
-
+            if (additional.Length != 0)
+            {
+                sources = (sources.Concat(new string[] { RestoreAdditionalProjectSources }).Concat(additional)).ToArray();
```

Comment:
I discussed this offline with @nkolev92 

I really don't like this method either, but the real fix would be to cache all the original nomination info instead of the packagespec, so a much bigger change.

This could cache an internal type that extends PackageSpec and adds a property bag where this extra info can live until the real spec is created. But that's still a hack somewhat.